### PR TITLE
PICARD-1411: Attempt to fix shared request counting

### PR DESCRIPTION
**Pull request by phw**
@pbrackin This is a pull requestb on top of your branch for https://github.com/metabrainz/picard-plugins/pull/181

The core issue I think is that special feature in this plugin where a single request kind of gets shared between multiple albums. In normal plugins we have a request per album. Here we have a single request that might be shared between multiple albums. This is tricky and easily causes albums to get stuck in loading if increasing and decreasing the request counters got out of sync.

In your branch the shared request loading was basically disabled. This worked in eliminating any , but actually could cause the album to be shown as completed but metadata getting updated silently later when the actual requests finished (I did not check, but possible without refreshing the UI).

I made some attempts to fix this. The key changes:

1. Reintroduce increasing the request counter when `item_id` is already in `self.albums`
2. Make sure `album` is never more than once in  `self.albums[item_id]` by making it a set
3. Separate the counters for open wikidata plugin requests from `album._requests`. This avoids the issue that the wikidata plugin relies on this counter to get zero to cleanup `self.albums[item_id]`, but other requests independent of wikidata might still be pending. In practice this luckily seldom happened since the wikidata is really slow and usually ended up as the last code touching the request counter
4. Lock the entire code in `musicbrainz_release_lookup`. I am not entirely sure what happened here, but this function touches the request counter multiple times (with each call of process_wikidata and at the end). Without this I got albums getting stuck with a request count of 1

I tested this with multiple releases. Most importantly multiple releases of the same artist, with reloading albums as soon as they got loaded. With this version it works, with the original wikidata plugin it causes albums getting stuck.

Can you do some tests on this iteration and merge it into your branch if it works for you?

### CODE REVIEW

File: plugins/wikidata/wikidata.py [hunk 1]
```diff
@@ -8,27 +8,32 @@
...
     def __init__(self):
         self.lock = threading.Lock()
         # Key: mbid, value: List of metadata entries to be updated when we have parsed everything
         self.requests = {}
-        
+
         # Key: mbid, value: List of items to track the number of outstanding requests
-        self.albums = {}
-        
+        self.albums = defaultdict(lambda: set())
```
Comment (42B) [inline plugins/wikidata/wikidata.py]: Why not just `defaultdict(set)`?

File: plugins/wikidata/wikidata.py [hunk 2]
```diff
@@ -8,27 +8,32 @@
...
+
+        # Key: album, value: Number of wikidata requests for this album
+        self.request_counts = defaultdict(lambda: 0)
```
Comment (42B) [inline plugins/wikidata/wikidata.py]: `defaultdict(int)`?

Comment (phw) [reply plugins/wikidata/wikidata.py]: Thanks, fixed

**Comment by pbrackin** [comment]
OK well I think I understand some of what you are saying, but I'm not sure its totally correct, or that you took an optimal approach with it.  But I went ahead and just put your updated version in place to try it.  At this point, it doesn't appear these additional changes are working for me, but I will let you interpret the results, since I am not totally sure how you are wanting it to work.

I took all the "letter G" artists in my library and dropped them into Picard.   This is around 20-25 artist folders which had previously been processed by Picard (have the MusicBrainz GUIDs), but I want to set better genre data for these & also make additional tweaks.  With these changes of yours in, all the files just sit in the left pane & nothing moves to the right & that is as far as it goes.  With just my changes, everything moves to the right, processing begins & ends for all albums, with the new genre data for most albums - just like always.  I have been assuming the latter is the desired result, but maybe I am wrong.  The latter does seem like a more desirable result to me though.

**Comment by pbrackin** [comment]
Disregard the last part about it not moving the albums over.  Its working now.  I am not sure why it didn't work the first time, but I went to run it from the command line to watch the logs and it suddenly started working.  I tested it a few times & it worked every time after that.

Once I got it working, I tested it both ways & also ran two instances side by side so I could compare the results.  The new changes you put it didn't seem to create any problems, but it also didn't seem to solve anything.  At least, for this test set, the results were identical.  Also I was able to switch releases for a given album in each code base & both worked fine.  Both processed the full set at around 2 minutes.   

So, I didn't see any benefit here, but perhaps my test cases didn't cover the scenario or edge case you are trying to address.  Perhaps you explain in more detail the test case that would be needed to show the benefit of your changes?

**Comment by phw** [comment]
> So, I didn't see any benefit here, but perhaps my test cases didn't cover the scenario or edge case you are trying to address. Perhaps you explain in more detail the test case that would be needed to show the benefit of your changes?

One way to test this is to load a couple of albums of the same artist at once, where only the artist itself provides any wikidata. Best way to simulate this is probably commenting out all but the request for albumartist_id in `process_track`.

With your original version the result will be:

1. The first album added will take a long time to load, since it is triggering the actual wikidata request
2. All following albums will be shown as loaded pretty quickly, because there is a request for this albumartist already pending and they don't start a new one
3. Once the first album finishes loading it will have an updated genre set. The other albums do not

With my patches the behavior changes. Still the first album triggers the request, but the other albums are also shown as loading as long as the first album loads. Once the first album is finished loading all other albums will also finish, and all albums will have the same genre set.

**Comment by pbrackin** [comment]
Alright I will see if I can reproduce this.  Sounds a bit far removed, but that is probably why I haven't seen it thus far. 

Give me a few days to find some time - its about to get super-maxed for me with the holidays here...

**Comment by pbrackin** [comment]
I have been testing this a little bit each day for the last few days - alternating with & without this commit. 
 For the most part, I get the exact same results with & without this commit.  So far I haven't been able to reproduce the special case you described with real-world tests & I haven't tried the forced method in-code yet.  

I did find one failure though:  After loading in a batch of 30 or so albums, I had removed one which had a duplicate file, so I could delete the duplicate file & re-add the folder (not sure this is the "right" approach for duplicates, but this is how I did it).  After I re-added it, it got stuck in loading.  Again, I removed it (while it was in the loading state) & re-added it again & got the same result.  I am going to try and reproduce this with & without this commit.  I will follow up.  I do have the log which shows the last wikidata print with 5 artist requests pending.  If you want to see the whole log I can post it somewhere if you want - just let me know.  But I will follow up if I can reproduce this or not & with additional test results.

UPDATE:  I was able to reproduce this.  It seems like the issue is not related to the deletion of the duplicate.  It seems to be for any removal & re-add.  I even tested with just a single album - loads fine the first time, but removal & re-add fails.  Only tested it on a few removal & re-adds, but it happened every time.  So, hopefully you can also easily reproduce it on your side.  Also, I tested with only my changes in place & the problem did not occur.

**Comment by phw** [comment]
I can reproduce. I actually agree with you that this is probably a case we should ignore. What the release is trying to do here, sharing request counting between multiple albums being loaded, is something that the current Picard API never intended and it is extremely difficult to get right and complicates the code.

What we probably should do on top of your changes:

1. Completely get rid of `self.albums[item_id]` as a list of albums. Currently your patch just rips out parts of the request sharing functionality, but leaves the rest of the code alone. Hence the `self.albums[item_id]` only ever can contain a single album. At least we don't need `self.albums[item_id]` to be an array, we could maybe even get rid of self.albums completely.
2. Run `album.update` on all affected albums after loading finishes. This will make sure the UI is refreshed, avoiding the issue that some data changes will not be shown

I took the freedom to merge my first commit here, which was just doing some general code cleanup, to your branch. But for the rest above I will not find the time soon, I would appreciate if you could take a look at that.

**Comment by phw** [comment]
I am not entirely sure what happened to this branch, somehow I messed up. But let's continue on the state of https://github.com/metabrainz/picard-plugins/pull/181

**Comment by pbrackin** [comment]
OK yeah I agree it seems overly complex & your insight on the intent of the API helps.  Part of the strategy you're outlining there - I was thinking along those lines for doing the second ticket, but we've come this far already now, so I guess might as well go the distance at this point...  I will pick up on [metabrainz#181](https://github.com/metabrainz/picard-plugins/pull/181)
